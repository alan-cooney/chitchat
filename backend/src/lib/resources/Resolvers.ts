import {
  DynamoDbDataSource,
  Assign,
  AttributeValues,
  KeyCondition,
  MappingTemplate,
  PrimaryKey,
} from "@aws-cdk/aws-appsync";
import { Construct } from "@aws-cdk/core";
import { join } from "path";

/**
 * Appsync Resolvers
 */
export default class Resolvers {
  constructor(scope: Construct, tableDS: DynamoDbDataSource) {
    tableDS.createResolver({
      typeName: "Mutation",
      fieldName: "addMessage",
      requestMappingTemplate: MappingTemplate.fromFile(
        join(__dirname, "../../velocity/addMessageRequest.vtl")
      ),
      responseMappingTemplate: MappingTemplate.dynamoDbResultItem(),
    });

    tableDS.createResolver({
      typeName: "Mutation",
      fieldName: "addParticipant",
      requestMappingTemplate: MappingTemplate.fromFile(
        join(__dirname, "../../velocity/addParticipantRequest.vtl")
      ),
      responseMappingTemplate: MappingTemplate.dynamoDbResultItem(),
    });

    /**
     * Queries
     */

    tableDS.createResolver({
      typeName: "Query",
      fieldName: "conversation",
      requestMappingTemplate: MappingTemplate.fromString(
        join(__dirname, "../../velocity/conversationRequest.vtl")
      ),
      responseMappingTemplate: MappingTemplate.fromFile(
        join(__dirname, "../../velocity/querySingleItemResponse.vtl")
      ),
    });

    /**
     * Nested queries
     */
    tableDS.createResolver({
      typeName: "User",
      fieldName: "conversations",
      requestMappingTemplate: MappingTemplate.fromString(
        join(__dirname, "../../velocity/userConversationsRequest.vtl")
      ),
      responseMappingTemplate: MappingTemplate.fromString(
        join(__dirname, "../../velocity/userConversationsResponse.vtl")
      ),
    });

    tableDS.createResolver({
      typeName: "Conversation",
      fieldName: "messages",
      requestMappingTemplate: MappingTemplate.fromString(
        join(__dirname, "../../velocity/userConversationsRequest.vtl")
      ),
      responseMappingTemplate: MappingTemplate.fromString(
        join(__dirname, "../../velocity/userConversationsResponse.vtl")
      ),
    });
  }
}
