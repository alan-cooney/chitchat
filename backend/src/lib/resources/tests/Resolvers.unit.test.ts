import "@aws-cdk/assert/jest";
import { SynthUtils } from "@aws-cdk/assert";
import { Stack } from "@aws-cdk/core";
import { GraphqlApi, Schema } from "@aws-cdk/aws-appsync";
import { AttributeType, Table } from "@aws-cdk/aws-dynamodb";
import Resolvers from "../Resolvers";

/**
 * Test stack
 */
const stack = new Stack();
const api = new GraphqlApi(stack, "api", {
  name: "test",
});
const table = new Table(stack, "table", {
  partitionKey: {
    name: "pk",
    type: AttributeType.STRING,
  },
});
const tableDS = api.addDynamoDbDataSource("tableDataSource", table);
new Resolvers(stack, tableDS);

it("creates resources matching the snapshot", () => {
  const cloudformation = SynthUtils.toCloudFormation(stack);
  expect(cloudformation).toMatchSnapshot();
});
