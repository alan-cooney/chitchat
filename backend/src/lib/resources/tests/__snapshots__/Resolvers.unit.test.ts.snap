// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`creates resources matching the snapshot 1`] = `
Object {
  "Resources": Object {
    "apiC8550315": Object {
      "Properties": Object {
        "AuthenticationType": "API_KEY",
        "Name": "test",
      },
      "Type": "AWS::AppSync::GraphQLApi",
    },
    "apiDefaultApiKey6AB8D7C4": Object {
      "DependsOn": Array [
        "apiSchema0EA92056",
      ],
      "Properties": Object {
        "ApiId": Object {
          "Fn::GetAtt": Array [
            "apiC8550315",
            "ApiId",
          ],
        },
      },
      "Type": "AWS::AppSync::ApiKey",
    },
    "apiSchema0EA92056": Object {
      "Properties": Object {
        "ApiId": Object {
          "Fn::GetAtt": Array [
            "apiC8550315",
            "ApiId",
          ],
        },
        "Definition": "",
      },
      "Type": "AWS::AppSync::GraphQLSchema",
    },
    "apitableDataSourceCFD298C7": Object {
      "Properties": Object {
        "ApiId": Object {
          "Fn::GetAtt": Array [
            "apiC8550315",
            "ApiId",
          ],
        },
        "DynamoDBConfig": Object {
          "AwsRegion": Object {
            "Ref": "AWS::Region",
          },
          "TableName": Object {
            "Ref": "table8235A42E",
          },
        },
        "Name": "tableDataSource",
        "ServiceRoleArn": Object {
          "Fn::GetAtt": Array [
            "apitableDataSourceServiceRole08EE8DB4",
            "Arn",
          ],
        },
        "Type": "AMAZON_DYNAMODB",
      },
      "Type": "AWS::AppSync::DataSource",
    },
    "apitableDataSourceMutationaddMessageResolverDB6D8CD5": Object {
      "DependsOn": Array [
        "apiSchema0EA92056",
        "apitableDataSourceCFD298C7",
      ],
      "Properties": Object {
        "ApiId": Object {
          "Fn::GetAtt": Array [
            "apiC8550315",
            "ApiId",
          ],
        },
        "DataSourceName": "tableDataSource",
        "FieldName": "addMessage",
        "Kind": "UNIT",
        "RequestMappingTemplate": "#set($input = $context.args.input)
#set($input.type = \\"MESSAGE\\")
#set($input.id = $util.autoId())

{
    \\"version\\": \\"2017-02-28\\",
    \\"operation\\": \\"PutItem\\",
    \\"key\\" : {
        \\"pk\\" : $util.dynamodb.toDynamoDBJson(\\"ORG#$ctx.args.input.organizationID#CONV#$ctx.args.input.id\\"),
        \\"sk\\" : $util.dynamodb.toDynamoDBJson(\\"CREATED#$util.time.nowEpochSeconds()#MSG#$input.id\\")
    },
    \\"attributeValues\\": $util.dynamodb.toMapValuesJson($input)
}",
        "ResponseMappingTemplate": "$util.toJson($ctx.result)",
        "TypeName": "Mutation",
      },
      "Type": "AWS::AppSync::Resolver",
    },
    "apitableDataSourceMutationaddParticipantResolverD635E16E": Object {
      "DependsOn": Array [
        "apiSchema0EA92056",
        "apitableDataSourceCFD298C7",
      ],
      "Properties": Object {
        "ApiId": Object {
          "Fn::GetAtt": Array [
            "apiC8550315",
            "ApiId",
          ],
        },
        "DataSourceName": "tableDataSource",
        "FieldName": "addParticipant",
        "Kind": "UNIT",
        "RequestMappingTemplate": "#set($input = $context.args.input)
#set($input.type = \\"PARTICIPANT\\")

{
    \\"version\\": \\"2017-02-28\\",
    \\"operation\\": \\"PutItem\\",
    \\"key\\" : {
        \\"pk\\" : $util.dynamodb.toDynamoDBJson(\\"ORG#$ctx.args.input.organizationID#USER#$ctx.args.input.userID\\"),
        \\"sk\\" : $util.dynamodb.toDynamoDBJson(\\"UPDATED#$util.time.nowEpochSeconds()#CONV#$ctx.args.input.conversationID\\")
    },
    \\"attributeValues\\": $util.dynamodb.toMapValuesJson($input)
}",
        "ResponseMappingTemplate": "$util.toJson($ctx.result)",
        "TypeName": "Mutation",
      },
      "Type": "AWS::AppSync::Resolver",
    },
    "apitableDataSourceMutationupdateUserResolver63DBAD87": Object {
      "DependsOn": Array [
        "apiSchema0EA92056",
        "apitableDataSourceCFD298C7",
      ],
      "Properties": Object {
        "ApiId": Object {
          "Fn::GetAtt": Array [
            "apiC8550315",
            "ApiId",
          ],
        },
        "DataSourceName": "tableDataSource",
        "FieldName": "updateUser",
        "Kind": "UNIT",
        "RequestMappingTemplate": "#set($input = $context.args.input)
#set($input.type = \\"USER\\")

{
    \\"version\\": \\"2017-02-28\\",
    \\"operation\\": \\"PutItem\\",
    \\"key\\" : {
        \\"pk\\" : $util.dynamodb.toDynamoDBJson(\\"ORG#$ctx.args.input.organizationID#USER#$ctx.args.input.id\\"),
        \\"sk\\" : $util.dynamodb.toDynamoDBJson(\\"USER#$ctx.args.input.id\\")
    },
    \\"attributeValues\\": $util.dynamodb.toMapValuesJson($input)
}",
        "ResponseMappingTemplate": "$util.toJson($ctx.result)",
        "TypeName": "Mutation",
      },
      "Type": "AWS::AppSync::Resolver",
    },
    "apitableDataSourceQueryconversationResolverEC1FB87A": Object {
      "DependsOn": Array [
        "apiSchema0EA92056",
        "apitableDataSourceCFD298C7",
      ],
      "Properties": Object {
        "ApiId": Object {
          "Fn::GetAtt": Array [
            "apiC8550315",
            "ApiId",
          ],
        },
        "DataSourceName": "tableDataSource",
        "FieldName": "conversation",
        "Kind": "UNIT",
        "RequestMappingTemplate": "/home/alan/repos/chitchat/backend/src/velocity/conversationRequest.vtl",
        "ResponseMappingTemplate": "## Return a single item (the first result) from a query
$util.toJson($ctx.result.items.get(0))",
        "TypeName": "Query",
      },
      "Type": "AWS::AppSync::Resolver",
    },
    "apitableDataSourceQueryuserResolver4A437C3B": Object {
      "DependsOn": Array [
        "apiSchema0EA92056",
        "apitableDataSourceCFD298C7",
      ],
      "Properties": Object {
        "ApiId": Object {
          "Fn::GetAtt": Array [
            "apiC8550315",
            "ApiId",
          ],
        },
        "DataSourceName": "tableDataSource",
        "FieldName": "user",
        "Kind": "UNIT",
        "RequestMappingTemplate": "{
    \\"version\\" : \\"2017-02-28\\",
    \\"operation\\" : \\"Query\\",
    \\"expression\\": \\"#pk = :pk\\",
    \\"expressionAttributeNames\\": {
        \\"#pk\\": \\"pk\\",
    },
    \\"expressionAttributeValues\\": {
        \\":pk\\": \\"ORG#$ctx.args.organizationID#USER#$ctx.args.id\\",
    }
}",
        "ResponseMappingTemplate": "## Return a single item (the first result) from a query
$util.toJson($ctx.result.items.get(0))",
        "TypeName": "Query",
      },
      "Type": "AWS::AppSync::Resolver",
    },
    "apitableDataSourceServiceRole08EE8DB4": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": "appsync.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
      },
      "Type": "AWS::IAM::Role",
    },
    "apitableDataSourceServiceRoleDefaultPolicy33FDDB21": Object {
      "Properties": Object {
        "PolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": Array [
                "dynamodb:BatchGetItem",
                "dynamodb:GetRecords",
                "dynamodb:GetShardIterator",
                "dynamodb:Query",
                "dynamodb:GetItem",
                "dynamodb:Scan",
                "dynamodb:ConditionCheckItem",
                "dynamodb:BatchWriteItem",
                "dynamodb:PutItem",
                "dynamodb:UpdateItem",
                "dynamodb:DeleteItem",
              ],
              "Effect": "Allow",
              "Resource": Array [
                Object {
                  "Fn::GetAtt": Array [
                    "table8235A42E",
                    "Arn",
                  ],
                },
                Object {
                  "Ref": "AWS::NoValue",
                },
              ],
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "apitableDataSourceServiceRoleDefaultPolicy33FDDB21",
        "Roles": Array [
          Object {
            "Ref": "apitableDataSourceServiceRole08EE8DB4",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "table8235A42E": Object {
      "DeletionPolicy": "Retain",
      "Properties": Object {
        "AttributeDefinitions": Array [
          Object {
            "AttributeName": "pk",
            "AttributeType": "S",
          },
        ],
        "KeySchema": Array [
          Object {
            "AttributeName": "pk",
            "KeyType": "HASH",
          },
        ],
        "ProvisionedThroughput": Object {
          "ReadCapacityUnits": 5,
          "WriteCapacityUnits": 5,
        },
      },
      "Type": "AWS::DynamoDB::Table",
      "UpdateReplacePolicy": "Retain",
    },
  },
}
`;
